File: packages/custom-gatekeeper/src/ports.ts
import type {
  CodeMatch,
  CommitRecord,
  IntakeSnapshot,
  IssueRecord,
  PullRequestRecord,
  RepositoryMetadata,
  RepositoryRef,
  ValidationDecision,
  VisualEvidence,
} from "./domain.js";

export interface SearchOptions {
  query: string;
  limit?: number;
}

export interface GitHubPort {
  getRepository(repository: RepositoryRef): Promise<RepositoryMetadata>;
  listTree(repository: RepositoryRef, ref: string, path?: string): Promise<string[]>;
  getFile(repository: RepositoryRef, ref: string, path: string): Promise<string>;
  searchCode(repository: RepositoryRef, options: SearchOptions): Promise<CodeMatch[]>;
  searchIssues(repository: RepositoryRef, options: SearchOptions): Promise<IssueRecord[]>;
  getIssue(repository: RepositoryRef, number: number): Promise<IssueRecord>;
  getIssueComments(repository: RepositoryRef, number: number): Promise<IssueRecord["comments"]>;
  searchPullRequests(repository: RepositoryRef, options: SearchOptions): Promise<PullRequestRecord[]>;
  getPullRequest(repository: RepositoryRef, number: number): Promise<PullRequestRecord>;
  getPullRequestDiff(repository: RepositoryRef, number: number): Promise<string>;
  getPullRequestDiscussion(repository: RepositoryRef, number: number): Promise<string[]>;
  searchCommits(repository: RepositoryRef, options: SearchOptions): Promise<CommitRecord[]>;
  getCommit(repository: RepositoryRef, sha: string): Promise<CommitRecord>;
  getCommitDiff(repository: RepositoryRef, sha: string): Promise<string>;
  createIssue(repository: RepositoryRef, input: { title: string; body: string; labels: string[] }): Promise<IssueRecord>;
  addLabels(repository: RepositoryRef, issueNumber: number, labels: string[]): Promise<void>;
}

export interface LanguageModelInput {
  request: string;
  repository: string;
  defaultBranchSha: string;
  evidence: Array<{ kind: string; source: string; summary: string; excerpt?: string }>;
  previousQuestion?: string;
  answer?: string;
  limits: { maxContextCharacters: number; maxOutputCharacters: number };
}

export interface LanguageModel {
  decide(input: LanguageModelInput): Promise<ValidationDecision>;
}

export interface BrowserPort {
  capture(url: string, options: { environment: string; width: number; height: number }): Promise<VisualEvidence>;
}

export interface IntakeStore {
  get(id: string): Promise<IntakeSnapshot | undefined>;
  put(intake: IntakeSnapshot): Promise<void>;
  list(): Promise<IntakeSnapshot[]>;
}

export interface Clock {
  now(): Date;
}

export interface IdGenerator {
  next(): string;
}

File: packages/agent-issue-core/src/ports.rs
use crate::model::*;
use async_trait::async_trait;

#[derive(Debug, Clone)]
pub struct SearchOptions {
    pub query: String,
    pub limit: usize,
}

#[async_trait(?Send)]
#[allow(dead_code)]
pub trait GitHubPort {
    async fn get_repository(
        &self,
        repository: &RepositoryRef,
    ) -> Result<RepositoryMetadata, String>;
    async fn list_tree(
        &self,
        repository: &RepositoryRef,
        reference: &str,
        path: Option<&str>,
    ) -> Result<Vec<String>, String>;
    async fn get_file(
        &self,
        repository: &RepositoryRef,
        reference: &str,
        path: &str,
    ) -> Result<String, String>;
    async fn search_code(
        &self,
        repository: &RepositoryRef,
        options: SearchOptions,
    ) -> Result<Vec<CodeMatch>, String>;
    async fn search_issues(
        &self,
        repository: &RepositoryRef,
        options: SearchOptions,
    ) -> Result<Vec<IssueRecord>, String>;
    async fn get_issue(
        &self,
        repository: &RepositoryRef,
        number: u64,
    ) -> Result<IssueRecord, String>;
    async fn get_issue_comments(
        &self,
        repository: &RepositoryRef,
        number: u64,
    ) -> Result<Vec<CommentRecord>, String>;
    async fn search_pull_requests(
        &self,
        repository: &RepositoryRef,
        options: SearchOptions,
    ) -> Result<Vec<PullRequestRecord>, String>;
    async fn get_pull_request(
        &self,
        repository: &RepositoryRef,
        number: u64,
    ) -> Result<PullRequestRecord, String>;
    async fn get_pull_request_diff(
        &self,
        repository: &RepositoryRef,
        number: u64,
    ) -> Result<String, String>;
    async fn get_pull_request_discussion(
        &self,
        repository: &RepositoryRef,
        number: u64,
    ) -> Result<Vec<String>, String>;
    async fn search_commits(
        &self,
        repository: &RepositoryRef,
        options: SearchOptions,
    ) -> Result<Vec<CommitRecord>, String>;
    async fn get_commit(
        &self,
        repository: &RepositoryRef,
        sha: &str,
    ) -> Result<CommitRecord, String>;
    async fn get_commit_diff(
        &self,
        repository: &RepositoryRef,
        sha: &str,
    ) -> Result<String, String>;
    async fn create_issue(
        &self,
        repository: &RepositoryRef,
        title: &str,
        body: &str,
        labels: &[String],
    ) -> Result<IssueRecord, String>;
    async fn add_labels(
        &self,
        repository: &RepositoryRef,
        issue_number: u64,
        labels: &[String],
    ) -> Result<(), String>;
}

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LanguageModelInput {
    pub request: String,
    pub repository: String,
    pub default_branch_sha: String,
    pub evidence: Vec<LanguageModelEvidence>,
    pub previous_question: Option<String>,
    pub answer: Option<String>,
    pub max_context_characters: usize,
    pub max_output_characters: usize,
}

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LanguageModelEvidence {
    pub kind: String,
    pub source: String,
    pub summary: String,
    pub excerpt: Option<String>,
}

#[async_trait(?Send)]
pub trait LanguageModel {
    async fn decide(&self, input: LanguageModelInput) -> Result<ValidationDecision, String>;
}

#[async_trait(?Send)]
pub trait BrowserPort {
    async fn capture(
        &self,
        url: &str,
        environment: &str,
        viewport: Viewport,
    ) -> Result<VisualEvidence, String>;
}

#[async_trait(?Send)]
pub trait IntakeStore {
    async fn get(&self, id: &str) -> Result<Option<IntakeSnapshot>, String>;
    async fn put(&self, intake: &IntakeSnapshot) -> Result<(), String>;
    async fn list(&self) -> Result<Vec<IntakeSnapshot>, String>;
}

pub trait Clock {
    fn now(&self) -> String;
}

pub trait IdGenerator {
    fn next(&self, owner_id: &str, request: &str) -> String;
}
